libraries/AP_Compass/AP_Compass.cpp

```diff
@@ -1163,20 +1189,30 @@ bool Compass::configured(uint8_t i)
         return false;
     }
 
-    // backup detected dev_id
-    int32_t dev_id_orig = _state[i].dev_id;
+    // exit immediately if dev_id hasn't been detected
+    if (_state[i].detected_dev_id == 0) {
+        return false;
+    }
+
+    // back up cached value of dev_id
+    int32_t dev_id_cache_value = _state[i].dev_id;
 
     // load dev_id from eeprom
     _state[i].dev_id.load();
 
-    // if different then the device has not been configured
-    if (_state[i].dev_id != dev_id_orig) {
-        // restore device id
-        _state[i].dev_id = dev_id_orig;
+    // if dev_id loaded from eeprom is different from detected dev id or dev_id loaded from eeprom is different from cached dev_id, compass is unconfigured
+    if (_state[i].dev_id != _state[i].detected_dev_id || _state[i].dev_id != dev_id_cache_value) {
```

Comment:
Does the second condition really matter? The only way I'm seeing for the second condition to be true is if it was the detection code changing it, in which case, the first condition is also true. Am I missing something?

Reply:
or if a user changed the ID by mavlink. I think the check is at worst harmless

Reply:
> or if a user changed the ID by mavlink.

No, if a user changed it via MAVLink, then it's saved, so the cache will be equal to the loaded value. If we cached the value on initialization then it could do something.

I agree it's harmless, but, unless I missed something, it is also useless.